File: starter/python/src/assetops/domain/work_order.py
"""Day 4 B3: aggregate root of AssetOps — HQ's promise to fix one root cause on one charger.

Invariant (R5): one open root-cause WO per (charger_id, fault_code).
The lookup "is there already an open WO for this fault?" needs the repository, so it happens outside
the aggregate (Day 5 B3 process manager + WorkOrderRepository.find_open_by_fault); the aggregate
records append_duplicate_report().
"""
from enum import Enum

from assetops.domain.events import DuplicateFaultReported, TechnicianAssigned, WorkOrderClosed, WorkOrderEvent, WorkOrderOpened  # noqa: F401


class WorkOrderStatus(str, Enum):
    OPEN = "Open"
    ASSIGNED = "Assigned"
    CLOSED = "Closed"


class WorkOrder:
    def __init__(self, work_order_id: str, charger_id: str, fault_code: str, connector_id: str | None, opened_at: str) -> None:
        self.work_order_id = work_order_id  # WO-2208
        self.charger_id = charger_id  # CP-A12
        self.fault_code = fault_code  # GroundFailure
        self.connector_id = connector_id  # CP-A12-2 (None = whole charger)
        self.opened_at = opened_at
        self.status = WorkOrderStatus.OPEN
        self.technician_id: str | None = None
        self.outcome: str | None = None
        self.reports: list[str] = []  # occurred_at of every report, first one included
        self._events: list[WorkOrderEvent] = []

    @classmethod
    def open(cls, *, work_order_id: str, charger_id: str, fault_code: str, at: str, connector_id: str | None = None) -> "WorkOrder":
        """Record the first report and a WorkOrderOpened event."""
        raise NotImplementedError("TODO R5: WorkOrder.open")

    @property
    def is_open(self) -> bool:
        return self.status is not WorkOrderStatus.CLOSED

    @property
    def duplicate_report_count(self) -> int:
        return len(self.reports) - 1

    def append_duplicate_report(self, at: str) -> None:
        """R5: only while open; append to reports, record DuplicateFaultReported."""
        raise NotImplementedError("TODO R5: append_duplicate_report")

    def assign(self, technician_id: str, at: str) -> None:
        raise NotImplementedError("TODO Day 5 B3: assign")

    def close(self, outcome: str, at: str) -> None:
        raise NotImplementedError("TODO Day 5 B3: close")

    def pull_events(self) -> list[WorkOrderEvent]:
        pulled, self._events = self._events, []
        return pulled

    def _record(self, event: WorkOrderEvent) -> None:
        self._events.append(event)

File: starter/node/src/assetops/domain/WorkOrder.ts
import type { WorkOrderEvent } from './events.ts';

// Day 4 B3: aggregate root of AssetOps — HQ's promise to fix one root cause on one charger.
// Invariant (R5): one open root-cause WO per (chargerId, faultCode).
// The lookup "is there already an open WO for this fault?" needs the repository, so it happens
// outside the aggregate (Day 5 B3 process manager); the aggregate records appendDuplicateReport().
export const WorkOrderStatus = { Open: 'Open', Assigned: 'Assigned', Closed: 'Closed' } as const;
export type WorkOrderStatus = (typeof WorkOrderStatus)[keyof typeof WorkOrderStatus];

export class WorkOrder {
  readonly workOrderId: string; // WO-2208
  readonly chargerId: string; // CP-A12
  readonly faultCode: string; // GroundFailure
  readonly connectorId: string | undefined; // CP-A12-2 (undefined = whole charger)
  readonly openedAt: string;
  readonly reports: string[] = []; // occurredAt of every report, first one included
  private _status: WorkOrderStatus = WorkOrderStatus.Open;
  private _technicianId?: string;
  private _outcome?: string;
  private events: WorkOrderEvent[] = [];

  private constructor(workOrderId: string, chargerId: string, faultCode: string, connectorId: string | undefined, openedAt: string) {
    this.workOrderId = workOrderId;
    this.chargerId = chargerId;
    this.faultCode = faultCode;
    this.connectorId = connectorId;
    this.openedAt = openedAt;
  }

  // Record the first report and a WorkOrderOpened event.
  static open(input: { workOrderId: string; chargerId: string; connectorId?: string; faultCode: string; at: string }): WorkOrder {
    throw new Error('TODO R5: WorkOrder.open');
  }

  get status(): WorkOrderStatus {
    return this._status;
  }
  get technicianId(): string | undefined {
    return this._technicianId;
  }
  get outcome(): string | undefined {
    return this._outcome;
  }
  get isOpen(): boolean {
    return this._status !== WorkOrderStatus.Closed;
  }
  get duplicateReportCount(): number {
    return this.reports.length - 1;
  }

  // R5: only while open; push to reports, record DuplicateFaultReported.
  appendDuplicateReport(at: string): void {
    throw new Error('TODO R5: appendDuplicateReport');
  }

  assign(technicianId: string, at: string): void {
    throw new Error('TODO Day 5 B3: assign');
  }

  close(outcome: string, at: string): void {
    throw new Error('TODO Day 5 B3: close');
  }

  pullEvents(): WorkOrderEvent[] {
    const pulled = this.events;
    this.events = [];
    return pulled;
  }

  private record(event: WorkOrderEvent): void {
    this.events.push(event);
  }
}
